README.md
# Beaker Cleanup Driver for GitLab Runners

<!-- vim-markdown-toc GFM -->

* [Description](#description)
* [Requirements](#requirements)
* [Setup](#setup)
  * [1. Installing the executor files](#1-installing-the-executor-files)
  * [2. Registering new Runner(s) using the configuration template file](#2-registering-new-runners-using-the-configuration-template-file)
* [Reference](#reference)
  * [Environment variables](#environment-variables)
  * [How the executor knows what VMs to clean up](#how-the-executor-knows-what-vms-to-clean-up)
  * [Troubleshooting](#troubleshooting)
    * [journald logs on syslog identifier `beaker-cleanup-driver`](#journald-logs-on-syslog-identifier-beaker-cleanup-driver)
    * [journald logs for `gitlab-runner` service](#journald-logs-for-gitlab-runner-service)
      * [Limitations](#limitations)

<!-- vim-markdown-toc -->

![](assets/beaker-cleanup-driver.gif)

## Description

**beaker-cleanup-driver** is a GitLab CI runner [custom executor][custom
executor] that properly cleans up each CI job's child processes and VMs―even if
the job failed, was cancelled, or timed out (probably).

1. Drop-in alternative to the shell executor
2. Tracks all processes created by a CI job and stops them after it terminates.
3. Includes special logic to clean up Virtual Box VMs and Vagrant sessions
   * (Using their respective tooling in order to properly reclaim their disk resources)

It is designed to run multi-node acceptance tests using Puppet's
[Beaker][beaker] while avoiding the shell executor's tendency to
accumulate hundreds of gigs of disk space and running VM processes, left
over from jobs that didn't terminate cleanly.


## Requirements

* Gitlab CI Runner 11+
* Linux host (tested on CentOS 7)
* bash

## Setup

### 1. Installing the executor files

As root, clone the contents of this repository into a directory on the runner,
and make sure that the `gitlab-runner` user can read and execute the directory
path and the `*.sh` scripts:

```sh

# We'll keep using this $CUSTOM_EXECUTOR_DIR in the next example, too
CUSTOM_EXECUTOR_DIR=/opt/simp/gitlab-runner/beaker-cleanup-driver
THIS_REPOSITORY_URL="<the url of this git repository>"

umask 0027
git clone "$THIS_REPOSITORY_URL" "$CUSTOM_EXECUTOR_DIR"

# Ensure that gitlab-runner can read and execute the scripts
chgrp -R gitlab-runner "$CUSTOM_EXECUTOR_DIR"
chmod -R g=u-w "$CUSTOM_EXECUTOR_DIR"
```

### 2. Registering new Runner(s) using the configuration template file

A Gitlab Runner [configuration template file][configuration template file]
is provided to simplify registration.

1. Tailor the template file `beaker-cleanup-driver.template.toml`

   * Make sure the `_exec` entries in the `[runners.custom]` section are under the
     path where you cloned the repository (`$CUSTOM_EXECUTOR_DIR`).
   * Tailor settings to taste (not everybody needs `log_level = 'debug')

2. Register a new GitLab Runner using the `--template-config` option:

```sh
gitlab-runner register \
  --name "$RUNNER_NAME" \
  --registration-token "$REGISTRATION_TOKEN" \
  --url "${CI_SERVER_URL:-https://gitlab.com/}" \
  --executor custom \
  --template-config "$CUSTOM_EXECUTOR_DIR/beaker-cleanup-driver.template.toml" \
  --tag-list "${RUNNER_TAG_LIST:-beaker}" \
  --non-interactive \
  --paused \
  --locked
```
For more information, see GitLab's documentation about [Registering Runners][registering runners]

## Reference


### Environment variables

You can set environment variables (in the `job_env` key of the executor's `config_exec.sh` script<sup>[[0]]</sup>)


| Environment Variable | Purpose                                                             | Default                      |
|----------------------|---------------------------------------------------------------------|------------------------------|
| `CI_RUNNER_USER`     | non-privileged build user                                           | `gitlab-runner`              |
| `CI_RUNNER_USER_DIR` | parent path for non-privileged build user's build/cache directories | `/var/lib/${CI_RUNNER_USER}` |

### How the executor knows what VMs to clean up

Beaker (and its VM orchestration) does not lend itself to using PID files, so
instead the custom executor "tags" every process it starts with a unique
environment variable:

1. The exec script in each stage sets an environment variable `$_CI_JOB_TAG` to
   a value that is unique to the job
2. Every runner script for the job executes with `$_CI_JOB_TAG` set to this same
   value.
3. When the `cleanup_exec` stage executes, all processes with `$_CI_JOB_TAG` set
   to that value will be cleaned up.
   - Running VirtualBox VMs will be shut down and deleted
   - `vagrant global-status --prune` will run if any VMs were shut down
   - after a short grace period, all remaining processes
4. The environment variables are found by grepping `/proc/*/environ` for the
   job's unique `$_CI_JOB_TAG` value
5. VMs are idenitified by grepping each `$_CI_JOB_TAG` pid's associated
   `/proc/$pid/cmdline` for VBoxHeadless

The technique of tagging processes for later cleanup using environment
variables (and identifying them via `/proc/*/environ`) was informed by the
discussion at https://serverfault.com/a/274613 .


### Troubleshooting

#### journald logs on syslog identifier `beaker-cleanup-driver`

The custom executor logs its activity to the Runner console output and to
journald under the syslog identifier `beaker-cleanup-driver`.

You can follow these messages via `journalctl`:

        journalctl -t beaker-cleanup-driver -f

While troubleshooting, you are likely to only be interested in messages from a
single job.  To see a specific job's progress, grep for the job's unique
`_CI_JOB_TAG` value (shown in the Runner console output at the beginning of
every exec stage and sub-stage):

        journalctl -t beaker-cleanup-driver -n 3000 -e  | egrep _CI_JOB_TAG=runner-1751173-project-17669670-concurrent-0-487152645 -A5 -B6 | less

#### journald logs for `gitlab-runner` service

The `gitlab-runner` unit's journald log can also be useful for correlating the
Runner's communication and setup events during troubleshooting.  Depending on
the gitlab-runner's `debug_level`, some executor messages (ONLY from the
`cleanup_exec` stage) will also appear in this log.

You can follow the unit's log via `journalctl`:

        journalctl -u gitlab-runner -f -n 3000

The gitlab-runner's log can have a lot of noise in addition to the executor
output To help filter the noise, you can grep the logs:

        journalctl -u gitlab-runner  -e -n 3000   | egrep '==|--' | less

To follow a specific job's progress, grep for the _CI_JOB_TAG value (shown in
the Runner console output at the beginning of every exec stage and sub-stage):

        journalctl -u gitlab-runner -e | grep _CI_JOB_TAG=runner-1751173-project-17669670-concurrent-0-485904412


##### Limitations

* The `gitlab-runner` unit's journald log only includes output from the
  `cleanup_exec` script (when `log_level` is `warn` or higher)
* There can be a signifcant delay before the messages show up at all.

[registering runners]: https://docs.gitlab.com/runner/register/
[configuration template file]: https://docs.gitlab.com/runner/register/#runners-configuration-template-file
[custom executor]: https://docs.gitlab.com/runner/executors/custom.html
[beaker]: https://github.com/puppetlabs/beaker
[0]: https://docs.gitlab.com/runner/executors/custom.html#config


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).



<!--
## [Unreleased]

### Added

### Changed

### Fixed

### Removed
-->

## [0.6.1] - 2021-03-04

### Fixed

- Fixed root cause of `pids: unbound variable` messages in base.sh
  - These were caused by referencing empty arrays, which is considered unset in
    Bash >= 4.1 < 4.4
  - For reference:
    | OS           | Bash version |
    | ------------ | ------------ |
    | EL7          | 4.2          |
    | EL8          | 4.4          |
    | Fedora >= 30 | 5.0+         |

### Changed

- Updated `ci_job_pids()` to only return tagged processes that have a
  cmdline and aren't the executing script (e.g., `cleanup_exec.sh`) or its
  functions
- Simplified logic in `ci_job stop()` and `ci_stop_tagged_jobs()`

### Removed

- Unused `ci_job_kill_procs()` function
- Unused `ci_job_cmdlines()` function

## [0.6.0] - 2021-03-03

### Added

- Documentation for environment variables

### Fixed

- Rewrote non-priv user stage script execution logic
  - Fixes failures on hardened systems where `/tmp` is mounted with `noexec`
  - Fixes security vulnerability that could be exploited to leak secrets to
    malicious code from other jobs/projects
  - For details, see note in `base.sh` comments about [gitlab-runner#4804]
- Fixed missing logs in some systems' journald by preferring `systemd-cat`
  when it is available
  - If `system-cat` is not available, system logging falls back to `logger`

### Removed

- `ci_job_ensure_user_can_access_script` function (no longer needed)

## [0.5.0] - 2020-09-04

### Fixed

- Fixed bug in ci_job_pids to set local `__CI_JOB_TAG`
- Fixed `line 56: _CI_JOB_TAG: unbound variable` message at the end of base.sh
- Fixed `line 84: _CI_JOB_TAG: unbound variable` message at the end of base.sh
- Fixed `line 133: _CI_JOB_TAG: unbound variable` message at the end of base.sh
- Script no longer terminates if `/proc/$pid/cmdline` doesn't exist
- Error checking and logging for various commands
- Fixed `logger` choking on strings beginning with `-`

### Added

- Various bash `local` touch-ups (to be extra careful)
- `pipe_warn` function to stream and log output from important commands

### Changed

- Consolidated stop() logic in `base.sh`
- Build path uses `CI_CONCURRENT_ID` instead of `CI_CONCURRENT_PROJECT_ID`
- Build path uses `CI_PROJECT_NAMESPACE/CI_PROJECT_NAME` instead of
  `CI_PROJECT_PATH_SLUG`
  - Ref: https://docs.gitlab.com/runner/best_practice/#build-directory
- `ci_job_pids()` gives `$1` precedence over `$_CI_JOB_TAG`


## [0.4.1] - 2020-06-17

### Fixed

- 0.4.0 cleanup routines didn't clean up VMs because they ran `vboxmanage` and
  `vagrant` as `root` instead of `$CI_RUNNER_USER`.

## [0.4.0] - 2020-06-10

### Added

- Builds are now executed as an *unprivileged* user (default: `gitlab-runner`)
- base.sh:
  - Added variables with defaults: `$CI_RUNNER_USER` and `$CI_RUNNER_USER_DIR`
    - These specify the non-privileged build user & build/cache parent path
  - Added `banner()` function

### Changed

- Custom executor bumped to version 0.4.0
- `*_exec.sh` scripts now report only their basenames without extensions
- config_exec.sh now sources `base.sh` for default variable values
- base.sh is now silent on stdout as it is being sourced

### Fixed

- Tagged release version has been advanced to match the custom executor (`0.4.0`)
- Safety checks in `ci_job()`


## [0.1.1] - 2020-05-27

### Added

- Travis pipeline w/shellcheck validation

### Changed

- Fixed `line 101: _CI_JOB_TAG: unbound variable` message at the end of base.sh
- Updated scripts to pass shellcheck validation


## [0.1.0] - 2020-04-12

### Added

- Initial release

[gitlab-runner#4804]: https://gitlab.com/gitlab-org/gitlab-runner/-/issues/4804

[0.1.0]: https://github.com/simp/gitlab-beaker-cleanup-driver/releases/tag/0.1.0
[0.1.1]: https://github.com/simp/gitlab-beaker-cleanup-driver/releases/tag/0.1.1
[0.4.0]: https://github.com/simp/gitlab-beaker-cleanup-driver/releases/tag/0.4.0
[0.4.1]: https://github.com/simp/gitlab-beaker-cleanup-driver/releases/tag/0.4.1
[0.5.0]: https://github.com/simp/gitlab-beaker-cleanup-driver/releases/tag/0.5.0
[0.6.0]: https://github.com/simp/gitlab-beaker-cleanup-driver/compare/0.5.0...0.6.0
[0.6.1]: https://github.com/simp/gitlab-beaker-cleanup-driver/compare/0.6.0...0.6.1
[Unreleased]: https://github.com/simp/gitlab-beaker-cleanup-driver/compare/0.6.1...HEAD


run_exec.sh
#!/bin/bash -Eeu
# ------------------------------------------------------------------------------
# GitLab Runner custom executor run script
# ------------------------------------------------------------------------------
# * Executes during each of run's sub-stages
# * Arguments:
#      $1 = Path to script GitLab wants to run for this sub-stage
#      $2 = sub-stage name
# * STDOUT and STDERR returned from this script will print to the job log
# * See: https://docs.gitlab.com/runner/executors/custom.html#run
# ------------------------------------------------------------------------------
script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
# shellcheck source=base.sh
source "${script_dir}/base.sh"
banner "$(basename --suffix='.sh' "${BASH_SOURCE[0]}")[ $2 ]"

notice "  +--- Run sub stage $2 ---+: $1 (_CI_JOB_TAG=$_CI_JOB_TAG)"

env > "/tmp/custom_executor.$2.env"
echo "$1" > "/tmp/custom_executor.script-path.$2.txt"
cat "$1" > "/tmp/custom_executor.script-content.$2.txt"


ci_job start "$1"


config_exec.sh
#!/bin/bash -Eeu
# ------------------------------------------------------------------------------
# GitLab Runner custom executor config script
# ------------------------------------------------------------------------------
# * Provides GitLab with information about the custom executor/"driver"
# * See: https://docs.gitlab.com/runner/executors/custom.html#config
# ------------------------------------------------------------------------------
script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
# shellcheck source=base.sh
source "${script_dir}/base.sh" >/dev/null 2>&1

cat << JSON
{
  "driver": {
    "name": "SIMP beaker cleanup driver",
    "version": "v0.6.1"
  },
  "builds_dir": "${CI_RUNNER_USER_DIR}/builds/${CUSTOM_ENV_CI_RUNNER_SHORT_TOKEN}/${CUSTOM_ENV_CI_CONCURRENT_ID}/${CUSTOM_ENV_CI_PROJECT_NAMESPACE}/${CUSTOM_ENV_CI_PROJECT_NAME}",
  "cache_dir":  "${CI_RUNNER_USER_DIR}/cache/${CUSTOM_ENV_CI_PROJECT_NAMESPACE}/${CUSTOM_ENV_CI_PROJECT_NAME}"
}
JSON


cleanup_exec.sh
#!/bin/bash -Eeu
# ------------------------------------------------------------------------------
# GitLab Runner custom executor cleanup script
# ------------------------------------------------------------------------------
# * Clean up any of the environments that might have been set up
# * This script is executed even if one of the previous stages failed
# * STDERR will be printed to logs when Runner `log_level = 'warn'` or higher
# * STDOUT will be printed to logs when Runner `log_level = 'debug'`
# * See: https://docs.gitlab.com/runner/executors/custom.html#cleanup
# ------------------------------------------------------------------------------
script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
# shellcheck source=base.sh
source "${script_dir}/base.sh"
banner "$(basename --suffix='.sh' "${BASH_SOURCE[0]}")"

ci_job stop



prepare_exec.sh
#!/bin/bash -Eeu
# ------------------------------------------------------------------------------
# GitLab Runner custom executor prepare script
# ------------------------------------------------------------------------------
# * Handles any custom actions necessary to set up the environment
# * See: https://docs.gitlab.com/runner/executors/custom.html#prepare
# ------------------------------------------------------------------------------
script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
# shellcheck source=base.sh
source "${script_dir}/base.sh"
banner "$(basename --suffix='.sh' "${BASH_SOURCE[0]}")"

#
# NOTE: There's nothing this stage needs to do (so far).
#

# Possibilities:
# ------------------------------------------------------------------------------
#   - Validate that RVM / the current Ruby Version is available to runner
#     - And if not: install it!
#   - Ensure that the cache is not messed up for this particular job
#     - What should we check for?
#     - In response to a problem, either;
#       a. Fix it (safely, without interfering with other jobs)
#       b. Fail the job immediately, with a helpful description that makes it
#          easier to figure out how to diagnose/fix the problem
# ------------------------------------------------------------------------------


base.sh
#!/bin/bash -Eeu
# ------------------------------------------------------------------------------
# Sets up the _CI_JOB_TAG environment variable and other common behaviors
# ------------------------------------------------------------------------------
# * /proc/*/environ-tagging idea informed by https://serverfault.com/a/274613
# ------------------------------------------------------------------------------

set -o pipefail

_CI_JOB_TAG="${_CI_JOB_TAG:-"runner-${CUSTOM_ENV_CI_RUNNER_ID}-project-${CUSTOM_ENV_CI_PROJECT_ID}-concurrent-${CUSTOM_ENV_CI_CONCURRENT_PROJECT_ID}-${CUSTOM_ENV_CI_JOB_ID}"}"

# Non-privileged user to execute the actual job script
CI_RUNNER_USER="${CI_RUNNER_USER:-gitlab-runner}"
_CI_RUNNER_USER_DIR_autodetect="$(getent passwd "$CI_RUNNER_USER" | cut -d: -f6)"
CI_RUNNER_USER_DIR="${CI_RUNNER_USER_DIR:-${_CI_RUNNER_USER_DIR_autodetect:-/var/lib/$CI_RUNNER_USER}}"
unset _CI_RUNNER_USER_DIR_autodetect

notice()
{
  echo "${@}"
  if command -v systemd-cat > /dev/null; then
    echo "${@}" | systemd-cat -t beaker-cleanup-driver -p info
  else
    logger -t beaker-cleanup-driver -- "${@}"
  fi
}

warn()
{
  >&2 echo "${@}"
  if command -v systemd-cat > /dev/null; then
    echo "${@}" | systemd-cat -t beaker-cleanup-driver -p warning
  else
    logger -t beaker-cleanup-driver -- "${@}"
  fi
}

pipe_notice()
{
   while IFS="" read -r data; do
     notice "$data"
   done
}

pipe_warn()
{
   while IFS="" read -r data; do
     warn "$data"
   done
}

banner()
{
  banner="======================================="
  notice "$(printf "\n\n%s\n\n    %s:  _CI_JOB_TAG=%s\n%s\n\n" "$banner" "${2:-${1:-$0}}" "$_CI_JOB_TAG" "$banner")"
}

# Return PIDs with a matching _CI_JOB_TAG env variable + cmdline
# Filters out the PIDs of the current script and function
#
# $1 = value of $_CI_JOB_TAG to match
ci_job_pids()
{
  local __CI_JOB_TAG="${1:-"${_CI_JOB_TAG:-NO_ARG_OR_ENV_VAR_GIVEN}"}"
  local -a pids=()
  # shellcheck disable=SC2153
  for pid in $(grep -l "\b_CI_JOB_TAG=$__CI_JOB_TAG\b" /proc/*/environ | cut -d/ -f3); do
    [ "$pid" == "$$" ] && continue
    [ "$pid" == "self" ] && continue
    [ "$pid" == "$BASHPID" ] && continue
    [ -f "/proc/$pid/cmdline" ] || continue
    ## warn "~~%%: $pid  $(cat "/proc/$pid/cmdline" || true)"
    pids+=($pid)
  done

  [ "${#pids[@]}" -gt 0 ] && echo "${pids[@]}"
}

# $@             = pids of VirtualBox VMs to stop
# $___ci_job_tag = outside-scope variable with _CI_JOB_TAG to kill
ci_job_stop_vbox()
{
  local -a pids
  if [ $# -gt 0 ]; then
    pids=("$@")
  else
    warn "== no pids to check"
    return 0
  fi

  local -a found_vbox_vms
  for pid in "${pids[@]}"; do
    [ -f "/proc/$pid/cmdline" ] || continue
    local -a pid_cmdline
    pid_cmdline=($(strings -1 < "/proc/$pid/cmdline")) || true
    if [[ "$(basename "${pid_cmdline[0]}")" = "VBoxHeadless" ]]; then
      local vbox_vm="${pid_cmdline[2]}"
      local vbox_uuid="${pid_cmdline[4]}"
      found_vbox_vms+=("$vbox_uuid")e

      warn "==== Powering off running VirtualBox VM '${vbox_vm}' (UUID='${vbox_uuid}') (pid='$pid')"
      pipe_warn < <(runuser -l "$CI_RUNNER_USER" -c "vboxmanage controlvm '$vbox_uuid' poweroff" 2>&1 ) || \
        warn "  !! poweroff failed for VM '${vbox_vm}'"

      warn "==== Unregistering VirtualBox VM '${vbox_vm}' (UUID='${vbox_uuid}') (pid='$pid')"
      pipe_warn < <(runuser -l "$CI_RUNNER_USER" -c "vboxmanage unregistervm '$vbox_uuid' --delete" 2>&1) || \
        warn "  !! unregistervm failed for VM '${vbox_vm}'"
    fi
  done

  if [ "${#found_vbox_vms[@]}" -gt 0 ]; then
    warn "____ Deleted ${#found_vbox_vms[@]} VirtualBox VMs (with _CI_JOB_TAG=${___ci_job_tag})"
    warn "==== Pruning any invalid vagrant environments"
    pipe_warn < <(runuser -l "$CI_RUNNER_USER" -c 'vagrant global-status --prune' 2>&1 || \
      echo "  !! 'vagrant global-status --prune' failed with exit code '$?'")
  else
    notice "____ No leftover running VirtualBox VMs were found (with _CI_JOB_TAG=${___ci_job_tag})"
  fi
}

# Ensure that all processes running with a matching value in $_CI_JOB_TAG are
# killed
#
# - Attempt to gracefully shut down VMs with the Beaker-generated Vagrantfile
# - Attempt to gracefully shut down all tagged VirtualBox processes
# - `kill` any tagged processes that are still running
#
#   $1: value of $_CI_JOB_TAG
ci_stop_tagged_jobs()
{
  local ___ci_job_tag="$1"
  local -a pids=($(ci_job_pids "$___ci_job_tag")) || true
  if [ "${#pids[@]}" -eq 0 ]; then
    warn "== no tagged pids to kill (with _CI_JOB_TAG=$___ci_job_tag)" && return 0
  fi

  notice "== Stopping any vagrant boxes running out of '$CUSTOM_ENV_CI_PROJECT_DIR/.vagrant/beaker_vagrant_files/default.yml'"
  pipe_warn < <(runuser -l "$CI_RUNNER_USER" -c 'vagrant global-status --prune' \
    | grep "$CUSTOM_ENV_CI_PROJECT_DIR/.vagrant/beaker_vagrant_files/default.yml" \
    | xargs -i runuser -l "$CI_RUNNER_USER" -c  "vagrant destroy -f {}" 2>&1 ) || warn "  !! exit-code: '$0'"

  notice "== Cleaning up any leftover VirtualBox VMs (with _CI_JOB_TAG=${___ci_job_tag})"
  ci_job_stop_vbox "${pids[@]}"
  sleep 8 # give post-VM processes a little time to die

  pids=($(ci_job_pids "$___ci_job_tag")) || true
  if [ "${#pids[@]}" -gt 0 ]; then
    notice "== killing ${#pids[@]} leftover pids (with _CI_JOB_TAG=$___ci_job_tag):"
    for pid in "${pids[@]}"; do
      warn "==   $pid    $(cat "/proc/$pid/cmdline" || true)"
    done
    kill "${pids[@]}"
  else
    warn "== no tagged pids left to kill (with _CI_JOB_TAG=$___ci_job_tag)"
  fi
}

# Start / stop a CI job
#   $1:
#     start = execute script, setting $_CI_JOB_TAG on all child processes
#     stop  = kill any processes
#   $2: path to executable stage script (provided by GitLab Runner)
ci_job()
{
  case "$1" in
  start)
    #
    # Situation:
    #
    # - GitLab Runner always writes custom executors' stage scripts under /tmp.
    # - These scripts (and their parent directories) are created as root/0700.
    # - The scripts' path, owner, and mode are not configurable.
    #
    #     https://gitlab.com/gitlab-org/gitlab-runner/-/issues/4804
    #
    # Problems:
    #
    # 1. Although the GitLab Runner writes the scripts as root/0700, we need
    #    to execute them with a non-privileged user (gitlab-runner). However,
    #    the restricted permisions prevent any non-privileged user from
    #    accessing the scripts (and their parent directories).
    #
    # 2. Naively relaxing stage scripts' owner/perms in order to grant access
    #    to the non-privileged user would introduce a small (but real) window
    #    of opportunity for malicious code from concurrently-executing
    #    pipelines to steal secrets from other projects.
    #
    # 3. /tmp is often mounted as `noexec` on hardened systems.  This prevents
    #    *any* user from executing the custom executor's stage scripts, even if
    #    they are permitted to access them.
    #
    # Solution:
    #
    #   The wily abomination below (inspired by a GL dev's comment in #4804)
    #
    script_content="$(cat "$2")"
    runuser -l "$CI_RUNNER_USER" -c "export _CI_JOB_TAG='$_CI_JOB_TAG'; $script_content"
    ;;
  stop)
    notice "== Stopping all related processes (with _CI_JOB_TAG=$_CI_JOB_TAG)"
    ci_stop_tagged_jobs "$_CI_JOB_TAG"
    notice "== Done stopping CI VMs + processes (with _CI_JOB_TAG=$_CI_JOB_TAG)"
    ;;
  esac
}
